File: iosApp/iosApp/ImageGalleryImpl.swift
import UIKit
import Photos
import SharedSDK


class ImageGalleryImpl:NSObject {

    let imageManager = PHImageManager.default()
    var fetchResult: PHFetchResult<PHAsset>?
    var selectedImages = [(UIImage, URL)]()


    func getImagess(completion: @escaping ([(UIImage, URL)]) -> Void) {
        let status = PHPhotoLibrary.authorizationStatus()
        if status == .authorized {
            self.fetchImages(completion: completion)
        } else if status == .denied || status == .restricted {
            completion([])
        } else {
            PHPhotoLibrary.requestAuthorization { (newStatus) in
                if newStatus == .authorized {
                    self.fetchImages(completion: completion)
                } else {
                    completion([])
                }
            }
        }
    }

    private func fetchImages(completion: @escaping ([(UIImage, URL)]) -> Void) {
        let fetchOptions = PHFetchOptions()
        fetchOptions.sortDescriptors = [NSSortDescriptor(key: "creationDate", ascending: false)]
        fetchResult = PHAsset.fetchAssets(with: .image, options: fetchOptions)

        guard let fetchResult = fetchResult else {
            completion([])
            return
        }

        let requestOptions = PHImageRequestOptions()
        requestOptions.deliveryMode = .fastFormat
        requestOptions.isSynchronous = true

        for i in 0..<fetchResult.count {
            imageManager.requestImage(for: fetchResult.object(at: i), targetSize: CGSize(width: 200, height: 200), contentMode: .aspectFill, options: requestOptions) { (image, info) in
                if let image = image, let url = info?["PHImageFileURLKey"] as? URL {
                    self.selectedImages.append((image, url))
                }
            }
        }

        completion(selectedImages)
    }
}

File: common/permission/camera/src/iosMain/kotlin/data/ImageGalleryImpl.kt
package data

import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import platform.CoreGraphics.CGSizeMake
import platform.Foundation.*
import platform.Photos.*
import platform.UIKit.UIImage

class ImageGalleryImpl : ImageGallery {
    private val imageManager = PHImageManager.defaultManager()
    private var selectedImages = mutableListOf<Pair<UIImage, NSURL>>()

    private fun getImages(completion: (List<Pair<UIImage, NSURL>>) -> Unit) {
        val status = PHPhotoLibrary.authorizationStatus()
        if (status == PHAuthorizationStatusAuthorized) {
            fetchImages(completion)
        } else if (status == PHAuthorizationStatusDenied || status == PHAuthorizationStatusRestricted) {
            completion(emptyList())
        } else {
            PHPhotoLibrary.requestAuthorization { newStatus ->
                if (newStatus == PHAuthorizationStatusAuthorized) {
                    fetchImages(completion)
                } else {
                    completion(emptyList())
                }
            }
        }
    }

    @OptIn(DelicateCoroutinesApi::class, ExperimentalForeignApi::class)
    private fun fetchImages(completion: (List<Pair<UIImage, NSURL>>) -> Unit) {
        val fetchOptions = PHFetchOptions()
        fetchOptions.sortDescriptors =
            listOf(NSSortDescriptor.sortDescriptorWithKey("creationDate", ascending = false))
        val fetchResult =
            PHAsset.fetchAssetsWithMediaType(PHAssetMediaTypeImage, options = fetchOptions)

        fetchResult.let { result ->
            val requestOptions = PHImageRequestOptions().apply {
                deliveryMode = PHImageRequestOptionsDeliveryModeFastFormat
                synchronous = true
            }

            GlobalScope.launch(Dispatchers.Default) {
                for (i in 0 until result.count.toInt()) {
                    imageManager.requestImageForAsset(
                        result.objectAtIndex(1.toULong()) as PHAsset,
                        targetSize = CGSizeMake(width = 200.0, height = 200.0),
                        contentMode = PHImageContentModeAspectFill,
                        options = requestOptions
                    ) { image, info ->
                        if (image != null && info != null) {
                            info["PHImageFileURLKey"]?.let { url ->
                                selectedImages.add(Pair(image, url as NSURL))
                            }
                        }
                    }
                }

                GlobalScope.launch(Dispatchers.Main) {
                    completion(selectedImages)
                }
            }
        }
    }

    override suspend fun getImages(): List<Image> {
        val image = mutableListOf<Image>()
        getImages {
            image.addAll(it)
        }
        return image
    }
}

private fun <E> MutableList<E>.addAll(elements: List<Pair<UIImage, NSURL>>) {

}